image-customize

```diff
@@ -82,6 +83,114 @@ class InstallAction(ActionBase):
         machine_instance.execute(f"{install_command} {package}", timeout=1800)
 
 
+class BuildAction(ActionBase):
+    '''Build and install distribution package(s) from dist tarball'''
+
+    quick = False
+    verbose = False
+
+    @staticmethod
+    def execute(machine_instance, tarball):
+        # upload the tarball
+        tarname = os.path.basename(tarball)
+        vm_tar = os.path.join("/var/tmp", tarname)
+        # looks like cockpit-foo-123.10.gdbde381.tar.xz
+        m = re.search(r'([^-]+)\.tar(\..*)?$', tarname)
+        if not m:
+            raise ValueError(f"cannot parse version number from tarball {tarname}")
+        version = m.group(1)
+        machine_instance.upload([tarball], vm_tar, relative_dir=".")
+
+        machine_instance.execute("rm -rf /var/tmp/build; mkdir -p /var/tmp/build")
+
+        # this will fail if neither is available -- exception is clear enough, this is a developer tool
+        out = machine_instance.execute("(which pbuilder || which mock || which pacman) 2>/dev/null")
+        if 'pbuilder' in out:
+            BuildAction.build_deb(machine_instance, vm_tar, version)
+        elif 'mock' in out:
+            BuildAction.build_rpm(machine_instance, vm_tar, version)
+        elif 'pacman' in out:
+            BuildAction.build_arch(machine_instance, vm_tar, version)
+        else:
+            raise NotImplementedError(f"unknown build platform: {out}")
+
+        # machine_instance.execute(f"{install_command} {package}", timeout=1800)
+
+    def build_deb(machine, vm_tar, version):
+        build_opts = 'nocheck' if BuildAction.quick else ''
+
+        # build source packge
+        machine.execute(f"""
+            set -eu
+            export DEB_BUILD_OPTIONS='{build_opts}'
+            tar -C /var/tmp/build -xf '{vm_tar}'
+            cd "$(ls -d /var/tmp/build/*)"
+            # find and copy debian packaging directory
+            cp -r "$(dirname $(find -path '*/debian/control'))" .
+            # create orig.tar link for building dsc
+            source=$(awk '/^Source: / {{ print $2 }}' debian/control)
+            ln -s '{vm_tar}' ../${{source}}_{version}.orig.tar.xz
+            # fill in the version number
+            sed -i '1 s/([^)-]*/({version}/' debian/changelog
+
+            dpkg-buildpackage -S -us -uc -nc""")
+
+        # build binary packages
+        machine.execute("cd /var/tmp/build; pbuilder build --buildresult . *.dsc", timeout=1800)
```

Comment:
Why isn't that part of the same script?

Reply:
I tried to group the machine.execute() blobs into "prepare source tree", "build source package", "build binary package", and "install packages", so that the `image-customize --verbose` output is a bit nicer to follow, and you get more targeted output on failures. It also makes it easier to copy&paste such a shell script into a VM for debugging this step.

The "prepare source tree" steps are rather involved, and all assume that /var/tmp/build is the cwd; these commands are also not independent from each other, so I think it's fine to keep them together as a single shell script.

Reply:
OK.